File: permute.js
var count = 0;
function permute(s, k) {
	if (k == 0) {
		count++;
		console.log(count + ' ' + s);
	} else {
		for (var i = k; i >= 0; i--) {
			var tmp;
			tmp = s[k]; s[k] = s[i]; s[i] = tmp;
			permute(s, k-1);
			tmp = s[k]; s[k] = s[i]; s[i] = tmp;
		}
	}
}

var a = ['A','B','C','D'];
permute(a, a.length-1);

File: permute.c
#include <stdio.h>
#include <string.h>

int count = 0;
void permute(char *s, int k) {
	if (k == 0) {
		count++;
		printf("%4d %s\n", count, s);
	} else {
		for (int i = k; i >= 0; i--) {
			char t;
			t = s[i]; s[i] = s[k]; s[k] = t;
			permute(s, k-1);
			t = s[i]; s[i] = s[k]; s[k] = t;
		}
	}
}

char s[26];
int main() {
	strcpy(s, "ABCDE");
	permute(s, strlen(s)-1);
	return 0;
}
